influxdata/tokenizer.go

```diff
@@ -0,0 +1,142 @@
+package influxdata
+
+import (
+	"io"
+)
+
+// minRead holds the minimum buffer size passed to io.Reader.Read.
+const minRead = 8192
+
+// NewTokenizer returns a tokenizer that splits the line-protocol text
+// inside buf. The text in buf represent a complete line (or multiple lines) - that
+// is, the tokenization will not work correctly if presented with a line
+// split at an arbitrary place.
+func NewTokenizerWithBytes(buf []byte) *Tokenizer {
+	return &Tokenizer{
+		buf:      buf,
+		complete: true,
+	}
+}
+
+func NewTokenizer(r io.Reader) *Tokenizer {
+	return &Tokenizer{
+		rd: r,
+	}
+}
+
+// Tokenizer implements low level parsing of a line-protocol message.
+type Tokenizer struct {
+	// rd holds the reader, if any. If there is no reader,
+	// complete will be true.
+	rd io.Reader
+
+	// buf holds read data.
+	buf []byte
+
+	// complete holds whether the data in buffer
+	// is known to be all the data that's available.
+	complete bool
+
+	// r holds the read position in buf.
+	r int
+
+	// err holds any non-EOF error that was returned from rd.
+	err error
+}
+
+// take returns the next slice of bytes within t.buf that are in the given set
+// starting at t.r + from, reading more data as needed.
+//
+// It does not update t.r.
+func (t *Tokenizer) take(set *byteSet, from int) []byte {
+	r1 := t.r + from
+outer:
+	for {
+		if t.isEmpty(r1) {
```

Comment:
Should this be `t.isEmpty(from)`? If so, could you add a test for this?

Reply:
Excellent catch! I'll add a test.

Reply:
`t.isEmpty(from)` is also wrong as it doesn't account for the incrementing! Should be `t.isEmpty(r1-t.r)` I believe.

Reply:
As it turns out, it's not that easy to write a test for, as the only down-side is that it might read a little extra in some cases, so I think I'll just fix it.